src/adapter/src/coord/command_handler.rs

```diff
@@ -683,6 +688,87 @@ impl Coordinator {
+                    // overestimation here by grabbing the ids of all indexes that are on ids
+                    // involved in the query. (We'd have to recursively follow view definitions,
+                    // similarly to `validate_timeline_context`.)
+                    // 2. For a peak, when the `least_valid_read` is later than the oracle's
+                    // `read_ts`, then the peak doesn't return before it completes at the chosen
+                    // timestamp. However, for a CRATE MATERIALIZED VIEW statement, it's not clear
+                    // whether we want to make it block until the chosen time. If it doesn't block,
+                    // then the initial refresh wouldn't be linearized with the CREATE MATERIALIZED
+                    // VIEW statement.
+                    //
+                    // Note: The Adapter is usually keeping a read hold of all objects at the oracle
+                    // read timestamp, so `least_valid_read` usually won't actually be later than
+                    // the oracle's `read_ts`. (see `Coordinator::advance_timelines`)
+                    //
+                    // Note 2: If we choose a timestamp here that is earlier than
+                    // `least_valid_read`, that is somewhat bad, but not catastrophic: The only
+                    // bad thing that happens is that we won't perform that refresh that was
+                    // specified to be at `mz_now()` (which is usually the initial refresh)
+                    // (similarly to how we don't perform refreshes that were specified to be in the
+                    // past).
+                } else {
+                    None
+                };
+
+                let owned_catalog = self.owned_catalog();
+                let catalog = owned_catalog.for_session(ctx.session());
+
+                purify_create_materialized_view_options(
+                    catalog,
+                    mz_now,
+                    &mut cmvs,
+                    &mut resolved_ids,
+                );
+
+                let purified_stmt =
+                    Statement::CreateMaterializedView(CreateMaterializedViewStatement::<Aug> {
+                        if_exists: cmvs.if_exists,
+                        name: cmvs.name,
+                        columns: cmvs.columns,
+                        in_cluster: cmvs.in_cluster,
+                        query: cmvs.query,
+                        with_options: cmvs.with_options,
+                    });
+
+                // (Purifying CreateMaterializedView doesn't happen async, so no need to send
+                // `Message::PurifiedStatementReady` here.)
+                match self.plan_statement(ctx.session(), purified_stmt, &params, &resolved_ids) {
+                    Ok(plan) => self.sequence_plan(ctx, plan, resolved_ids).await,
+                    Err(e) => ctx.retire(Err(e)),
+                }
+            }
```

Comment:
I created a separate function for two reasons:

1. I didn't want to unconditionally pass the current time, to avoid the cost of querying the timestamp oracle. @jkosh44 [says](https://github.com/MaterializeInc/materialize/pull/23870#discussion_r1433012494) calling the timestamp oracle is costly. What do you think, @benesch? If we decide that we don't want to unconditionally call the timestamp oracle, then I don't see a way to avoid extending the SQL crate's public API: Even if I were to somehow merge my purification function into the already existing purification function, I'd have to first call `materialized_view_option_contains_temporal` to decide whether to call the timestamp oracle before the call of the merged purification function.

2. The existing purification function is async, but I didn't want to perform my purification async, because I was afraid of the chosen timestamp becoming invalid by the time the async function comes back. I'm not actually sure whether this could happen. I thought it could because `message_purified_statement_ready` has a comment "DROP CONNECTION may have sneaked in", which made me scared that other things may also sneak in that invalidate the timestamp. Is this fear valid or am I misunderstanding something?

Reply:
Ok, fair enough. Thinking about it a bit more, it's the inconsistency in the purification API that bothers me more than the expansion of the API surface area—i.e., that `purify_statement` _seems_ to be a thing that can handle all statements, but in fact it can only handle `CREATE SOURCE`, `ALTER SOURCE`, and `CREATE SINK`, and now there will be a totally separate path for `CREATE MATERIALIZED VIEW`.

An alternate way of slicing this would be to have four separate functions:

  * `async fn purify_create_source()`
  * `async fn purify_alter_source()`
  * `async fn purify_create_sink()`
  * `fn purify_create_materialized_view()`

Then it would be more clear that the adapter needs to handle each one of these functions slightly differently. As it stands, `message_purified_statement_ready` is pretty tangled up because handling the results of `purify_{create|alter}_source` has almost nothing in common with handling `purify_create_sink`.

I did a quick exploration of what would be required to do the above refactoring and ... it's grim. At least 1-2 days of work, and that's optimistic. So, let's put a pin in this entire line of thinking and forget I said anything.

----

> I was afraid of the chosen timestamp becoming invalid by the time the async function comes back. I'm not actually sure whether this could happen.

There can be an arbitrary delay, for sure, between when purification starts and when the coordinator handles the "purification ready" message. What it means for the chosen timestamp to become invalid here is your wheelhouse, though! Is the idea that selecting `mz_now()` as the oracle read timestamp is "safe" because the adapter always has read holds on every object at `mz_now()` and there is no gap between observing the oracle read timestamp and planning/installing the materialized view in the catalog (at which point it will install its own read holds for the next refresh)? If so, you're right to be worried about using `async` here, and having a sync function that never yields control of the coordinator's main loop is definitely the way to go.

Reply:
> I did a quick exploration of what would be required to do the above refactoring and ... it's grim. At least 1-2 days of work, and that's optimistic. So, let's put a pin in this entire line of thinking and forget I said anything.

Ok. I've added a code comment linking your above message.

> There can be an arbitrary delay, for sure, between when purification starts and when the coordinator handles the "purification ready" message. What it means for the chosen timestamp to become invalid here is your wheelhouse, though! Is the idea that selecting mz_now() as the oracle read timestamp is "safe" because the adapter always has read holds on every object at mz_now() and there is no gap between observing the oracle read timestamp and planning/installing the materialized view in the catalog (at which point it will install its own read holds for the next refresh)?

Exactly.

> If so, you're right to be worried about using async here, and having a sync function that never yields control of the coordinator's main loop is definitely the way to go.

Ok, I've extended the doc comment of `purify_create_materialized_view_options` with the above reasoning of why it should not be async.

This is the commit ("Comments around purification"): https://github.com/MaterializeInc/materialize/pull/23870/commits/e6c152c7221bc483ad845e16ddb27c247f554f7a